Documentation (#17)
spacemeshos/go-spacecraft: docs/devnet.md

## Deploying a Devnet

This document explains the flow of deploying a devnet and web services.

### Authentication

> First you need to set up authentication to GCP. Download spacecraft-id-f8aeb7a6d337.json from 1password and place it in ~ and then set the following ENVs:

Reviewer: ```suggestion
First you need to set up authentication to GCP. 

```
We can keep the private info out.

```
export SPACECRAFT_GCP_PROJECT=spacecraft-id
export SPACECRAFT_GCP_LOCATION=us-central1
export SPACECRAFT_GCP_ZONE=us-central1-b
export GOOGLE_APPLICATION_CREDENTIALS=~/spacecraft-id-f8aeb7a6d337.json
```

And then run the below command to activate gcloud to use the service account:

```
gcloud auth activate-service-account --key-file ~/spacecraft-id-f8aeb7a6d337.json
```

### Set Cloudflare API Token

Set cloudflare API token in ENV so that all the services can be exposed publicly: 

```
export SPACECRAFT_CLOUDFLARE_API_TOKEN=
```

> You can find the token in 1password

Reviewer: We can keep the private info out.

```

If you want to change go-spacemesh config then you can change that in ./artifacts/devnet/miner/config.json and if you want to change infrastructure config then you can change that in ./artifacts/devnet/config.json. You can also run “./gospacecraft --help” to know all the available CLI options.

The process takes 3-5 minutes to finish and then at the end Kibana and Pyroscope URLs will be printed. 

You can find the newly deployed k8s cluster here https://console.cloud.google.com/kubernetes/list?authuser=1&project=spacecraft-id

### Configure Kubectl

To point kubectl to the deployed k8s cluster run the below command:

```
gcloud container clusters get-credentials devnet --region us-central1 --project spacecraft-id
```

### Metrics

Promethues stack is deployed by default and metrics is enabled in managed miners. Here is the credentials for grafana:

Username: admin
Password: prom-operator

### Smapp Config

When you deploy a network using go-spacecraft it creates and uploads a config file to GCP storage. You can access it using https://console.cloud.google.com/storage/browser/spacecraft-data/devnet-archive

### Slack Alerts

> To enable slack alerts you need to set the following environment variables:

Reviewer: Plz link this somehow to the "Monitoring and Alerts" clause in `README.md`.

```
export SPACECRAFT_SLACK_CHANNEL_ID=
export SPACECRAFT_SLACK_TOKEN=
```

> You can find the slack token in 1password and channel ID is the channel where you want to publish the alerts to.

Reviewer: ```suggestion
Channel ID is the channel where you want to publish the alerts to.

```
We can keep the private info out.

### Deploy Web Services

> To deploy web services first you need to download tls.key and tls.crt files from 1password and place it in ./artifacts/ws directory.

Reviewer: ```suggestion
To deploy web services first you first need to place the secret `tls.key` and `tls.crt` files in ./artifacts/ws directory.

```
Also, plz give some more info about these files. How were they created, why are they needed, etc.

Author: Added more information about it
